src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -854,25 +855,55 @@ supplyCredits ::
 supplyCredits c levels =
     V.iforM_ levels $ \_i (Level mr _rs) ->
     -- let !ln = i + 1 in
-    let !cr = creditsForMerge mr in
-    supplyMergeCredits (ceiling (fromIntegral c * cr)) mr
+    let !c' = scaleCreditsForMerge mr c in
+    supplyMergeCredits c' mr
 
-creditsForMerge :: MergingRun m h -> Rational
-creditsForMerge SingleRun{}                             = 0
-creditsForMerge (MergingRun LevelLevelling _ _)         = 1 + 4
-creditsForMerge (MergingRun LevelTiering (NumRuns n) _) = fromIntegral n / 4
+-- | 'Credit's scaled based on the merge requirements for merging runs. See
+-- 'scaleCreditsForMerge'.
+newtype ScaledCredits = ScaledCredits Int
 
-{-# SPECIALISE supplyMergeCredits :: Credit -> MergingRun IO h -> IO () #-}
+-- | Scale a number of credits to a number of merge steps to be performed, based
+-- on the merging run.
+--
+-- Initially, 1 update supplies 1 credit. However, since merging runs have
+-- different numbers of input runs/entries, we may have to a more or less
+-- merging work than 1 merge step for each credit.
+scaleCreditsForMerge :: MergingRun m h -> Credit -> ScaledCredits
+-- A single run is a trivially completed merge, so it requires no credits.
+scaleCreditsForMerge SingleRun{} _ = ScaledCredits 0
+-- A levelling merge has 1 input run and one resident run, which is (up to) 4x
+-- bigger than the others. It needs to be completed before another run comes in.
+--
+-- TODO: this is currently assuming a naive worst case, where the resident run
+-- is as large as it can be for the current level. We probably have enough
+-- information available here to lower the worst-case upper bound by looking at
+-- the sizes of the input runs. As as result, merge work would/could be more
+-- evenly distributed over time when the resident run is smaller than the worst
+-- case.
+scaleCreditsForMerge (MergingRun LevelLevelling _ _) c =
+    ScaledCredits (c * (1 + 4))
+-- A tiering merge has 5 runs at most (once could be held back to merged again)
+-- and must be completed before the level is full (once 4 more runs come in).
+scaleCreditsForMerge (MergingRun LevelTiering (NumRuns n) _) c
+  | y * 4 >= x = ScaledCredits y
+  | otherwise  = ScaledCredits (y + 1)
+  where
+    x :: Int
+    !x = c * n
+    y :: Int
+    !y = x `div` 4
```

Comment:
I think I see. So the bit:
```
  | y * 4 >= x = ScaledCredits y
  | otherwise  = ScaledCredits (y + 1)
```
is to round up the division.

And really what we want is `ceiling (c * n / 4)`.

So isn't the more direct way to do that, to do rounding up in the division. For example, to divide by 4 rounded up we can do ((a + 3) `div` 4).

Sanity test
```
ghci> and [ ((x + 3) `div` 4) == ceiling (toRational x / 4) | x <- [0..1000 :: Int] ]
True
```
So we could have:
```
scaleCreditsForMerge (MergingRun LevelTiering (NumRuns n) _) c =
    ScaledCredits ((c * n + 3) `div` 4)
    -- same as division rounding up: ceiling (c * n / 4)
```

Reply:
Yes, sounds simpler